### ISSUE TITLE
Add public getters/setters for mbedtls_md5_context, mbedtls_sha1_context etc

### ISSUE THREAD

**Issue by graeme-walker** [Mbed-TLS/mbedtls#5198]
### Suggested enhancement

Allow full access to 'state' and 'total' fields in mbedtls_md5_context, mbedtls_sha1_context and mbedtls_sha256_context.

### Justification

It is sometimes necessary to extract and store the intermediate state of a hash function and use that state later on to complete the hash once the full input is known.

Allowing public access to 'state' and 'total' fields in mbedtls_(hash)_context or providing equivalent getter and setter functions for the intermediate state would enable this usage pattern with mbedtls v3. Access to the 'buffer' field is not necessary since extraction of the intermediate state would only be valid  when the number of bytes adedd via mbedtls_(hash)_update() is some multiple of  the hash function's block size.

A specific use case is precomputing the HMAC intermediate state for CRAM-MD5 authentication so that passwords do not have to be stored as plaintext.

HMAC is described in RFC-2104, which notes that "the idea is that the intermediate results of the compression function on the B-byte blocks (K XOR ipad) and (K XOR opad) can be precomputed only once at the time of generation of the key K. These intermediate results are stored and then used to initialize the IV of H each time that a message needs to be authenticated."

RFC-2195 also notes that "the use of the techniques described .. for precomputation of intermediate results make it possible to avoid explicit cleartext storage of the shared secret on the server system by instead storing the intermediate results".

**Comment by mpg** [comment]
Hi @graeme-walker and thanks for getting in touch!

Sorry if I'm missing something obvious, but it seems like the use case you have in mind is what the existing `mbedtls_sha1_clone()` (and similar) functions do. If those functions don't meet your needs, can you explain why?

(And indeed the application to HMAC is something that had been on my mind for a long time, I think it was even in one of our internal trackers at some point.)

**Comment by graeme-walker** [comment]
The clone functions are not sufficient because they don't allow the intermediate state to be stored on disk, reloaded at some later date, and then have more data applied in order to complete the hash. The crux of the problem is that the hardware and software environment might be different between the storing and the reloading, so a simple memcpy or whatever is not sufficient.

I my case I support OpenSSL and mbedtls so a HMAC 'password' from an OpenSSL MD5_CTX needs to be loadable into an mbedtls_md5_context and vice versa. It also needs to work across versions so that a password generated by mbedtls2.x can be reloaded in ten years time by an application built with mbedtls4.x.

Everything works fine at the moment because MD5_CTX has Nh, Nl, A, B, C and D data members which map exactly to mbedtls_md5 total, state[0], state[1], state[2] and state[3] and all members are publically accessible.

Data hiding is clearly a good thing, but it seems to me that the state of the hashing function is not really an implementation detail that needs hiding because it fundamental to the published algorithms.

Thanks for your consideration -- Graeme

**Comment by gilles-peskine-arm** [comment]
> seems to me that the state of the hashing function is not really an implementation detail that needs hiding because it fundamental to the published algorithms

The internal state of a hashing function _is_ an implementation detail, even if there isn't much room for variation with commnly used algorithms. Even in Mbed TLS 2.x, the format of `mbedtls_md5_context` and friends was only guaranteed if you were not using an accelerator (`MBEDTLS_MD5_ALT`).

> the intermediate state to be stored on disk, reloaded at some later date, and then have more data applied in order to complete the hash

As you're seeing, this is difficult to implement if you don't have full control over the implementation of the hash. I guess it's too late in your case, but you can generally get the same security properties by chaining hashes: instead of calculating `hash(old_string || new_string)` and storing the intermediate state after `old_string`, calculate `hash(hash(old_string) || new_string)` and store `hash(old_string)`.

> It also needs to work across versions so that a password generated by mbedtls2.x can be reloaded in ten years time by an application built with mbedtls4.x.

Note that it's plausible that Mbed TLS 4.0 might drop MD5. It depends on how much legacy usage of MD5 remains at that point. We dropped MD2 and MD4 in Mbed TLS 3.0.

-----

Over the course of Mbed TLS 3.x, we intend to transition low-level crypto APIs to be closer matches for the PSA crypto API. As a consequence, I don't think we will be exposing low-level crypto data structures any more than they are now.

The PSA API includes a pair of functions that would fulfill your purpose: [`psa_hash_suspend`](https://armmbed.github.io/mbed-crypto/html/api/ops/hashes.html?highlight=psa_hash_suspend#c.psa_hash_suspend) and [`psa_hash_resume`](https://armmbed.github.io/mbed-crypto/html/api/ops/hashes.html?highlight=psa_hash_suspend#c.psa_hash_resume). They expose the hash state indirectly, allowing the implementation to convert from its internal representation as necessary. [We have a backlog task to implement those functions](https://github.com/ARMmbed/mbedtls/issues/3274), which is currently very low priority due to a lack of demand. Would you be willing to contribute an implementatoin of `psa_hash_{suspend,resume}`?

**Comment by graeme-walker** [comment]
> I don't think we will be exposing low-level crypto data structures any more than they are now.

That's fair enough; I was just worried that a blanket policy of encapsulation was clobbering one use case unnecessarily.

My project uses mbedtls mainly for TLS, but I also use the hashing API for CRAM-MD5 (etc) authentication, with the option of storing 'encrypted' passwords as HMAC IVs. If necessary I can bring the hashing code 'in-house' and/or fully deprecate the storing of encrypted passwords.

I am certainly open to contributing to the relevant psa functions. I note that there is no constraint in the descriptions about using complete blocks. The HMAC use case only requires complete blocks and it can therefore just extract the current data size and the (eg) four chaining variables. Is it feasible to extract the state 'mid-round'?

**Comment by gilles-peskine-arm** [comment]
I'm closing this issue because we are concentrating our efforts on the PSA crypto APIs rather than the legacy APIs. We are not going to add public/getter setters for low-level hash contexts. Adding a suspend/resume functionality to hashes is an API break since it requires additional work from alternative implementations (`MBEDTLS_xxx_ALT`). In the next major version of Mbed TLS, we intend to deprecate alternative implementations in favor of PSA drivers. This functionality should instead be implemented via `psa_hash_{suspend,resume}` (https://github.com/ARMmbed/mbedtls/issues/3274) (and the corresponding driver entry points if applicable).
